File: library/src/main/java/com/github/alexjlockwood/kyrie/PropertyTimeline.java
package com.github.alexjlockwood.kyrie;

import android.support.annotation.IntRange;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

final class PropertyTimeline {
  private final List<Property<?>> properties = new ArrayList<>();
  private final Property.Listener listener =
      new Property.Listener() {
        @Override
        public void onCurrentPlayTimeChanged(@NonNull Property<?> property) {
          drawable.invalidateSelf();
        }
      };

  @NonNull private final KyrieDrawable drawable;
  private long totalDuration;

  public PropertyTimeline(KyrieDrawable drawable) {
    this.drawable = drawable;
  }

  @NonNull
  public <V> Property<V> registerAnimatableProperty(List<Animation<?, V>> animations) {
    final Property<V> property = new Property<>(animations);
    properties.add(property);
    property.addListener(listener);
    if (totalDuration != Animation.INFINITE) {
      final long currTotalDuration = property.getTotalDuration();
      if (currTotalDuration == Animation.INFINITE) {
        totalDuration = Animation.INFINITE;
      } else {
        totalDuration = Math.max(currTotalDuration, totalDuration);
      }
    }
    return property;
  }

  public void setCurrentPlayTime(@IntRange(from = 0) long currentPlayTime) {
    for (int i = 0, size = properties.size(); i < size; i++) {
      properties.get(i).setCurrentPlayTime(currentPlayTime);
    }
  }

  public long getTotalDuration() {
    return totalDuration;
  }
}

File: kyrie/src/main/java/com/github/alexjlockwood/kyrie/PropertyTimeline.kt
package com.github.alexjlockwood.kyrie

import androidx.annotation.IntRange

import java.util.ArrayList

internal class PropertyTimeline(private val drawable: KyrieDrawable) {

    private val properties = ArrayList<Property<*>>()
    private val listener = object : Property.Listener {
        override fun onCurrentPlayTimeChanged(property: Property<*>) {
            drawable.invalidateSelf()
        }
    }

    var totalDuration: Long = 0
        private set

    fun <V> registerAnimatableProperty(animations: List<Animation<*, V>>): Property<V> {
        val property = Property(animations)
        properties.add(property)
        property.addListener(listener)
        if (totalDuration != Animation.INFINITE) {
            val currTotalDuration = property.totalDuration
            totalDuration = if (currTotalDuration == Animation.INFINITE) {
                Animation.INFINITE
            } else {
                Math.max(currTotalDuration, totalDuration)
            }
        }
        return property
    }

    fun setCurrentPlayTime(@IntRange(from = 0) currentPlayTime: Long) {
        properties.forEach { it.setCurrentPlayTime(currentPlayTime) }
    }
}
